crates/walrus-service/src/client/daemon/routes.rs

```diff
@@ -210,33 +301,43 @@ pub(super) async fn get_blob<T: WalrusReadClient>(
         Ok(consistency_check) => consistency_check,
         Err(error) => return error.into_response(),
     };
-    match client.read_blob(&blob_id, consistency_check).await {
-        Ok(blob) => {
-            tracing::debug!("successfully retrieved blob");
-            let mut response = (StatusCode::OK, blob).into_response();
-            let headers = response.headers_mut();
-            populate_response_headers_from_request(
-                request_method,
-                &request_headers,
-                &blob_id.to_string(),
-                headers,
-            );
-            response
-        }
-        Err(error) => {
-            let error = GetBlobError::from(error);
 
-            match &error {
-                GetBlobError::BlobNotFound => {
-                    tracing::debug!(?blob_id, "the requested blob ID does not exist")
-                }
-                GetBlobError::Internal(error) => tracing::error!(?error, "error retrieving blob"),
-                _ => (),
+    let mut response = if let Some(range_header) = request_headers.get(RANGE) {
+        let (start, length) = match parse_range_header(range_header) {
+            Ok(v) => v,
+            Err(error) => {
+                tracing::debug!(?error, "invalid byte range header");
+                return error.to_response();
             }
+        };
 
-            error.to_response()
+        // TODO(WAL-1107): support streaming the response sliver by sliver.
+        match client.read_byte_range(&blob_id, start, length).await {
+            Ok(blob) => (StatusCode::PARTIAL_CONTENT, blob).into_response(),
+            Err(error) => {
+                tracing::debug!(?error, "failed to read byte range");
+                return GetBlobError::from(error).to_response();
+            }
```

Comment:
`client.read_byte_range` could still return an error if start_index+length are not in the blob range. Right now, it seems like an error from `read_byte_range` i.e. `InvalidByteRange` is getting mapped back to internal error (default branch in the GetBlobError::from enum mapping) when it should be 400/416.